File: FlutterMobileapp/lib/models/todo.dart
/// Todo Model
/// Data model for todo items
class Todo {
  final String id;
  final String userId;
  final String title;
  final String? description;
  final TodoStatus status;
  final TodoPriority priority;
  final DateTime? dueDate;
  final DateTime createdAt;
  final DateTime updatedAt;

  Todo({
    required this.id,
    required this.userId,
    required this.title,
    this.description,
    required this.status,
    required this.priority,
    this.dueDate,
    required this.createdAt,
    required this.updatedAt,
  });

  /// Create from JSON
  factory Todo.fromJson(Map<String, dynamic> json) {
    return Todo(
      id: json['id'] ?? '',
      userId: json['userId'] ?? '',
      title: json['title'] ?? '',
      description: json['description'],
      status: TodoStatus.fromString(json['status'] ?? 'pending'),
      priority: TodoPriority.fromString(json['priority'] ?? 'medium'),
      dueDate: json['dueDate'] != null
          ? DateTime.tryParse(json['dueDate'].toString())
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.tryParse(json['updatedAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }

  /// Convert to JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'title': title,
      'description': description,
      'status': status.value,
      'priority': priority.value,
      'dueDate': dueDate?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  /// Copy with updated values
  Todo copyWith({
    String? title,
    String? description,
    TodoStatus? status,
    TodoPriority? priority,
    DateTime? dueDate,
  }) {
    return Todo(
      id: id,
      userId: userId,
      title: title ?? this.title,
      description: description ?? this.description,
      status: status ?? this.status,
      priority: priority ?? this.priority,
      dueDate: dueDate ?? this.dueDate,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }

  /// Check if todo is overdue
  bool get isOverdue {
    if (dueDate == null || status == TodoStatus.completed) return false;
    return DateTime.now().isAfter(dueDate!);
  }
}

/// Todo Status enum
enum TodoStatus {
  pending('pending', 'Pending'),
  inProgress('in_progress', 'In Progress'),
  completed('completed', 'Completed');

  final String value;
  final String label;

  const TodoStatus(this.value, this.label);

  static TodoStatus fromString(String value) {
    return TodoStatus.values.firstWhere(
      (s) => s.value == value,
      orElse: () => TodoStatus.pending,
    );
  }
}

/// Todo Priority enum
enum TodoPriority {
  low('low', 'Low'),
  medium('medium', 'Medium'),
  high('high', 'High');

  final String value;
  final String label;

  const TodoPriority(this.value, this.label);

  static TodoPriority fromString(String value) {
    return TodoPriority.values.firstWhere(
      (p) => p.value == value,
      orElse: () => TodoPriority.medium,
    );
  }
}

File: backend/src/models/Todo.js
/**
 * Todo Model
 * Firestore schema for todo documents
 */
const { collections, admin } = require('../config/firebase');
const { v4: uuidv4 } = require('uuid');

/**
 * Todo Schema:
 * {
 *   id: string,
 *   userId: string,
 *   title: string,
 *   description: string | null,
 *   status: 'pending' | 'in_progress' | 'completed',
 *   priority: 'low' | 'medium' | 'high',
 *   dueDate: timestamp | null,
 *   createdAt: timestamp,
 *   updatedAt: timestamp
 * }
 */

class TodoModel {
    static collection = collections.todos;

    /**
     * Create a new todo
     */
    static async create(todoData) {
        const id = uuidv4();
        const now = admin.firestore.FieldValue.serverTimestamp();

        const todo = {
            id,
            userId: todoData.userId,
            title: todoData.title,
            description: todoData.description || null,
            status: todoData.status || 'pending',
            priority: todoData.priority || 'medium',
            dueDate: todoData.dueDate ? new Date(todoData.dueDate) : null,
            createdAt: now,
            updatedAt: now,
        };

        await this.collection.doc(id).set(todo);
        return { ...todo, id, createdAt: new Date(), updatedAt: new Date() };
    }

    /**
     * Find todo by ID
     */
    static async findById(id) {
        const doc = await this.collection.doc(id).get();
        if (!doc.exists) return null;
        return this._formatTodo(doc.data());
    }

    /**
     * Find todos by user ID with optional filters
     */
    static async findByUserId(userId, options = {}) {
        const { limit = 50, status, priority, sortBy = 'createdAt', sortOrder = 'desc' } = options;

        let query = this.collection.where('userId', '==', userId);

        if (status) {
            query = query.where('status', '==', status);
        }

        if (priority) {
            query = query.where('priority', '==', priority);
        }

        // query = query.orderBy(sortBy, sortOrder).limit(limit);

        const snapshot = await query.get();
        let todos = snapshot.docs.map(doc => this._formatTodo(doc.data()));

        // In-memory sort and limit to avoid missing index errors
        todos.sort((a, b) => {
            const valA = new Date(a[sortBy] || 0).getTime();
            const valB = new Date(b[sortBy] || 0).getTime();
            return sortOrder === 'desc' ? valB - valA : valA - valB;
        });

        return todos.slice(0, limit);
    }

    /**
     * Update todo
     */
    static async update(id, updates) {
        const updateData = {
            ...updates,
            updatedAt: admin.firestore.FieldValue.serverTimestamp(),
        };

        // Handle dueDate conversion
        if (updates.dueDate !== undefined) {
            updateData.dueDate = updates.dueDate ? new Date(updates.dueDate) : null;
        }

        await this.collection.doc(id).update(updateData);
        return this.findById(id);
    }

    /**
     * Toggle todo status
     */
    static async toggleStatus(id) {
        const todo = await this.findById(id);
        if (!todo) return null;

        const statusFlow = {
            'pending': 'in_progress',
            'in_progress': 'completed',
            'completed': 'pending',
        };

        const newStatus = statusFlow[todo.status] || 'pending';

        return this.update(id, { status: newStatus });
    }

    /**
     * Delete todo
     */
    static async delete(id) {
        await this.collection.doc(id).delete();
        return true;
    }

    /**
     * Get todo statistics for user
     */
    static async getStatsByUserId(userId) {
        const snapshot = await this.collection
            .where('userId', '==', userId)
            .get();

        const stats = {
            total: 0,
            pending: 0,
            in_progress: 0,
            completed: 0,
            high_priority: 0,
        };

        snapshot.docs.forEach(doc => {
            const todo = doc.data();
            stats.total++;
            stats[todo.status] = (stats[todo.status] || 0) + 1;
            if (todo.priority === 'high') {
                stats.high_priority++;
            }
        });

        return stats;
    }

    /**
     * Format todo timestamps
     */
    static _formatTodo(todo) {
        if (!todo) return null;
        return {
            ...todo,
            createdAt: todo.createdAt?.toDate?.() || todo.createdAt,
            updatedAt: todo.updatedAt?.toDate?.() || todo.updatedAt,
            dueDate: todo.dueDate?.toDate?.() || todo.dueDate,
        };
    }
}

module.exports = TodoModel;
